executor/merge_join.go

```diff
@@ -0,0 +1,419 @@
+		return false, errors.Errorf("Invalid arguments: Empty arguments detected.")
+	}
+	rb.stmtCtx = rb.ctx.GetSessionVars().StmtCtx
+	var err error
+	rb.peekedRow, err = rb.nextRow()
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+
+	return rb.peekedRow == nil, nil
+}
+
+func (rb *rowBlockIterator) nextRow() (*Row, error) {
+	for {
+		row, err := rb.reader.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if row == nil {
+			rb.reader.Close()
+			return nil, nil
+		}
+		if rb.filter != nil {
+			var matched bool
+			matched, err = expression.EvalBool(rb.filter, row.Data, rb.ctx)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if !matched {
+				continue
+			}
+		}
+		return row, nil
+	}
+}
+
+func (rb *rowBlockIterator) nextBlock() ([]*Row, error) {
+	var err error
+	peekedRow := rb.peekedRow
+	var curRow *Row
+	if peekedRow == nil {
+		return nil, nil
+	}
+	rowCache := []*Row{peekedRow}
+	for {
+		curRow, err = rb.nextRow()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if curRow == nil {
+			rb.peekedRow = nil
+			return rowCache, nil
+		}
+		compareResult, err := compareKeys(rb.stmtCtx, curRow, rb.joinKeys, rb.peekedRow, rb.joinKeys)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		if compareResult == 0 {
+			rowCache = append(rowCache, curRow)
+		} else {
+			rb.peekedRow = curRow
+			return rowCache, nil
+		}
+	}
+}
+
+// Close implements the Executor Close interface.
+func (e *MergeJoinExec) Close() error {
+	e.finished.Store(true)
+
+	e.prepared = false
+	e.cursor = 0
+	e.outputBuf = nil
+
+	lErr := e.leftRowBlock.reader.Close()
+	rErr := e.rightRowBlock.reader.Close()
+	if lErr != nil {
+		return lErr
+	}
+	return rErr
+}
+
+// Schema implements the Executor Schema interface.
+func (e *MergeJoinExec) Schema() *expression.Schema {
+	return e.schema
+}
+
+func compareKeys(stmtCtx *variable.StatementContext,
+	leftRow *Row, leftKeys []*expression.Column,
+	rightRow *Row, rightKeys []*expression.Column) (int, error) {
+	for i, leftKey := range leftKeys {
+		lVal, err := leftKey.Eval(leftRow.Data)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		rVal, err := rightKeys[i].Eval(rightRow.Data)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		ret, err := lVal.CompareDatum(stmtCtx, rVal)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+
+		if ret != 0 {
+			return ret, nil
+		}
+	}
+	return 0, nil
+}
+
+func (e *MergeJoinExec) outputJoinRow(leftRow *Row, rightRow *Row) error {
+	var err error
+	joinedRow := makeJoinRow(leftRow, rightRow)
+	if e.otherFilter != nil {
+		matched, err := expression.EvalBool(e.otherFilter, joinedRow.Data, e.ctx)
+		if err != nil && matched {
+			e.outputBuf = append(e.outputBuf, joinedRow)
+		}
+	} else {
+		e.outputBuf = append(e.outputBuf, joinedRow)
+	}
+	return err
+}
+
+func (e *MergeJoinExec) computeJoin() (bool, error) {
+	e.outputBuf = make([]*Row, 0, rowBufferSize)
```

Comment:
We can use `e.outputBuf = e.outpuBuf[0:0:rowBufferSize]` to avoid memory allocation.